api/v1alpha1/executionhook_types.go

```diff
@@ -0,0 +1,180 @@
+	// If PodContainerNamesList is specified, this field will not be used.
+	// +optional
+	PodContainerSelector *PodContainerSelector `json:"podContainerSelector,omitempty" protobuf:"bytes,2,opt,name=podContainerSelector"`
+}
+
+// PodContainerNames lists the containers the ExecutionHook should be executed
+// on in a Pod.
+type PodContainerNames struct {
+	// This field is required
+	PodName string `json:"podName" protobuf:"bytes,1,opt,name=podName"`
+
+	// +optional
+	ContainerNames []string `json:"containerNames,omitempty" protobuf:"bytes,2,rep,name=containerNames"`
+}
+
+// PodContainerSelector defines the selector and containers the ExecutionHook
+// should be executed on.
+type PodContainerSelector struct {
+	// PodSelector specifies a label query over a set of pods.
+	// +optional
+	PodSelector *metav1.LabelSelector `json:"podSelector,omitempty" protobuf:"bytes,1,opt,name=podSelector"`
+
+	// If specified, controller only select the containers that are listed from the selected pods based on PodSelector.
+	// Otherwise, all containers of the pods will be selected
+	// +optional
+	ContainerList []string `json:"containerList,omitempty" protobuf:"bytes,2,rep,name=containerList"`
+}
+
+// ExecutionHookStatus defines the observed state of ExecutionHook
+type ExecutionHookStatus struct {
+	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// This is a list of ContainerExecutionHookStatus, with each status representing
+	// information about how hook is executed in a container, including pod name,
+	// container name, ActionTimestamp, ActionSucceed, etc.
+	// +optional
+	HookStatuses []ContainerExecutionHookStatus `json:"hookStatuses,omitempty" protobuf:"bytes,1,rep,name=hookStatuses"`
+}
+
+// ContainerExecutionHookStatus represents the current state of a hook for a specific
+// container in a pod
+type ContainerExecutionHookStatus struct {
+	// This field is required
+	PodName string `json:"podName" protobuf:"bytes,1,opt,name=podName"`
+
+	// This field is required
+	ContainerName string `json:"containerName" protobuf:"bytes,2,opt,name=containerName"`
+
+	// If not set, it is nil, indicating Action has not started
+	// If set, it means Action has started at the specified time
+	// +optional
+	Timestamp *metav1.Time `json:"actionTimestamp,omitempty" protobuf:"bytes,3,opt,name=actionTimestamp"`
+
+	// Succeed is set to true when the action is executed in the container successfully.
+	// It will be set to false if the action cannot be executed successfully after
+	// ActionTimeoutSeconds passes.
+	// +optional
+	Succeed *bool `json:"actionSucceed,omitempty" protobuf:"varint,4,opt,name=actionSucceed"`
+
+	// The last error encountered when executing the action. The hook controller might
+	// update this field each time it retries the execution.
+	// +optional
+	Error *HookError `json:"error,omitempty" protobuf:"bytes,5,opt,name=error"`
```

Comment:
There can be only one error? That is not an unreasonable assumption. However, this pattern seems to be similar to the Conditions pattern implemented by most built in types. Is there a reason to go with a singular error over Status.Conditions?

Reply:
The expectation is that the error represents the error condition encountered executing the hook action on a specific container in a pod. So there is a 1:1 mapping between where the hook action was executed (Pod,Container) and an error condition, if any. For that reason, using a singular error was thought of as a better fit. However, using `Status.Conditions` would be more appropriate if/ when we aggregate the errors into the status of the `ExecutionHook` object.

Reply:
I think we were given recommendations not to use Conditions during the API review, although I couldn't remember the reason behind that.   @thockin @jingxu97 do you want to comment on this?